File: OLD_MarioKart/MarioKart/TWGL/TWGLNode.m
//
//  RWTModel.m
//  HelloOpenGL
//
//  Created by Main Account on 3/18/14.
//  Copyright (c) 2014 Razeware LLC. All rights reserved.
//

#import "TWGLNode.h"
#import "TWGLShadersReference.h"
#import <OpenGLES/ES2/glext.h>
#import "TWGLOBJLoader.h"

@implementation TWGLNode {
    char *_name;
    GLuint _vao;
    GLuint _vertexBuffer;
    GLuint _indexBuffer;
    unsigned int _vertexCount;
    unsigned int _indexCount;
    TWGLShadersReference *_shader;
//    GLMmodel *model;
}

- (instancetype)initWithName:(char *)name shader:(TWGLShadersReference *)shader
{
    self = [self initWithName:name shader:shader vertices:NULL vertexCount:0 inidices:NULL indexCount:0];
    if (self) {
    }
    return self;
}

- (instancetype)initWithName:(char *)name shader:(TWGLShadersReference *)shader modelNamed:(NSString *)modelNamed {
    if (self = [super init]) {
//        model = [TWGLOBJLoader loadOBJ:modelNamed];
        _name = name;
        _shader = shader;
        self.position = GLKVector3Make(0, 0, 0);
        self.rotationX = 0;
        self.rotationY = 0;
        self.rotationZ = 0;
        self.scale = 1.0;
        self.children = [NSMutableArray array];
//        glmCreateArrays(model, GLM_NONE);

//        glGenVertexArraysOES(1, &_vao);
//        glBindVertexArrayOES(_vao);
//        
//        // Generate vertex buffer
//        
//        glGenBuffers(1, &_vertexBuffer);
//        glBindBuffer(GL_ARRAY_BUFFER, _vertexBuffer);
//        glBufferData(GL_ARRAY_BUFFER, vertexCount * sizeof(GLfloat), vertices, GL_STATIC_DRAW);
//        
//        // Generate index buffer
//        glGenBuffers(1, &_indexBuffer);
//        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, _indexBuffer);
//        glBufferData(GL_ELEMENT_ARRAY_BUFFER, indexCount * sizeof(GLubyte), indices, GL_STATIC_DRAW);
//        
//        // Enable vertex attributes
//        glEnableVertexAttribArray(TWGLVertexAttribPosition);
//        glVertexAttribPointer(TWGLVertexAttribPosition, 3, GL_FLOAT, GL_FALSE, 3*sizeof(GLfloat), 0);
        //        glEnableVertexAttribArray(TWGLVertexAttribColor);
        //        glVertexAttribPointer(TWGLVertexAttribColor, 4, GL_FLOAT, GL_FALSE, sizeof(TWGLVertex), (const GLvoid *) offsetof(TWGLVertex, Color));
        //        glEnableVertexAttribArray(TWGLVertexAttribTexCoord);
        //        glVertexAttribPointer(TWGLVertexAttribTexCoord, 2, GL_FLOAT, GL_FALSE, sizeof(TWGLVertex), (const GLvoid *) offsetof(TWGLVertex, TexCoord));
        //        glEnableVertexAttribArray(TWGLVertexAttribNormal);
        //        glVertexAttribPointer(TWGLVertexAttribNormal, 3, GL_FLOAT, GL_FALSE, sizeof(TWGLVertex), (const GLvoid *) offsetof(TWGLVertex, Normal));
        
//        glBindVertexArrayOES(0);
//        glBindBuffer(GL_ARRAY_BUFFER, 0);
//        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
        
    }
    return self;
}

- (instancetype)initWithName:(char *)name shader:(TWGLShadersReference *)shader vertices:(GLfloat *)vertices vertexCount:(unsigned int)vertexCount inidices:(GLubyte *)indices indexCount:(unsigned int)indexCount {
    
    if (self = [super init]) {
        
        _name = name;
        _vertexCount = vertexCount;
        _indexCount = indexCount;
        _shader = shader;
        self.position = GLKVector3Make(0, 0, 0);
        self.rotationX = 0;
        self.rotationY = 0;
        self.rotationZ = 0;
        self.scale = 1.0;
        self.children = [NSMutableArray array];
        
        glGenVertexArraysOES(1, &_vao);
        glBindVertexArrayOES(_vao);
        
        // Generate vertex buffer

        glGenBuffers(1, &_vertexBuffer);
        glBindBuffer(GL_ARRAY_BUFFER, _vertexBuffer);
        glBufferData(GL_ARRAY_BUFFER, vertexCount * sizeof(GLfloat), vertices, GL_STATIC_DRAW);
        
        // Generate index buffer
        glGenBuffers(1, &_indexBuffer);
        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, _indexBuffer);
        glBufferData(GL_ELEMENT_ARRAY_BUFFER, indexCount * sizeof(GLubyte), indices, GL_STATIC_DRAW);
        
        // Enable vertex attributes
        glEnableVertexAttribArray(TWGLVertexAttribPosition);
        glVertexAttribPointer(TWGLVertexAttribPosition, 3, GL_FLOAT, GL_FALSE, 3*sizeof(GLfloat), 0);
//        glEnableVertexAttribArray(TWGLVertexAttribColor);
//        glVertexAttribPointer(TWGLVertexAttribColor, 4, GL_FLOAT, GL_FALSE, sizeof(TWGLVertex), (const GLvoid *) offsetof(TWGLVertex, Color));
//        glEnableVertexAttribArray(TWGLVertexAttribTexCoord);
//        glVertexAttribPointer(TWGLVertexAttribTexCoord, 2, GL_FLOAT, GL_FALSE, sizeof(TWGLVertex), (const GLvoid *) offsetof(TWGLVertex, TexCoord));
//        glEnableVertexAttribArray(TWGLVertexAttribNormal);
//        glVertexAttribPointer(TWGLVertexAttribNormal, 3, GL_FLOAT, GL_FALSE, sizeof(TWGLVertex), (const GLvoid *) offsetof(TWGLVertex, Normal));
        
        glBindVertexArrayOES(0);
        glBindBuffer(GL_ARRAY_BUFFER, 0);
        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
        
    }
    return self;
}

- (GLKMatrix4)modelMatrix {
    GLKMatrix4 modelMatrix = GLKMatrix4Identity;
    modelMatrix = GLKMatrix4Translate(modelMatrix, self.position.x, self.position.y, self.position.z);
    modelMatrix = GLKMatrix4Rotate(modelMatrix, self.rotationX, 1, 0, 0);
    modelMatrix = GLKMatrix4Rotate(modelMatrix, self.rotationY, 0, 1, 0);
    modelMatrix = GLKMatrix4Rotate(modelMatrix, self.rotationZ, 0, 0, 1);
    modelMatrix = GLKMatrix4Scale(modelMatrix, self.scale, self.scale, self.scale);
    return modelMatrix;
}

- (void)renderWithParentModelViewMatrix:(GLKMatrix4)parentModelViewMatrix {
    
    GLKMatrix4 modelViewMatrix = GLKMatrix4Multiply(parentModelViewMatrix, [self modelMatrix]);
    for (TWGLNode *child in self.children) {
        [child renderWithParentModelViewMatrix:modelViewMatrix];
    }
    
    _shader.modelViewMatrix = modelViewMatrix;
    _shader.texture = self.texture;
    [_shader prepareToDraw];
    

//    if (model) {
//        glmDrawArrays(model, 0);
//    } else {
        glBindVertexArrayOES(_vao);
        glDrawElements(GL_TRIANGLES, _indexCount, GL_UNSIGNED_BYTE, 0);
        glBindVertexArrayOES(0);
//    }
}

- (void)updateWithDelta:(NSTimeInterval)dt {
    for (TWGLNode *child in self.children) {
        [child updateWithDelta:dt];
    }
}

- (void)loadTexture:(NSString *)filename {
    NSError *error;
    NSString *path = [[NSBundle mainBundle] pathForResource:filename ofType:nil];
    
    NSDictionary *options = @{ GLKTextureLoaderOriginBottomLeft: @YES };
    GLKTextureInfo *info = [GLKTextureLoader textureWithContentsOfFile:path options:options error:&error];
    if (info == nil) {
        NSLog(@"Error loading file: %@", error.localizedDescription);
    } else {
        self.texture = info.name;
    }
}

@end

File: OLDMarioKart/MarioKart/TWGL/TWGLNode.swift
//
//  TWGLNode.swift
//  MarioKart
//
//  Created by Txai Wieser on 11/3/15.
//  Copyright © 2015 Txai Wieser. All rights reserved.
//

import GLKit
import OpenGLES

class TWGLNode {
    let name:String
    var vao = GLuint()
    var vertexBuffer = GLuint()
    var indexBuffer = GLuint()

//    var vertexPointBuffer = GLuint()
//    var vertexUVBuffer = GLuint()
    var vertexCount:Int = 0
    weak var shader:TWGLShaderReference?
    
    var texture = GLuint()
    //    var size:GLKVector3 = 0
    
    
    var position:(x:Float, y:Float, z:Float) = (0, 0, 0)
    var rotation:(x:Float, y:Float, z:Float) = (0, 0, 0)
    var scale:Float = 1.0
    var matColor:(r:Float, g:Float, b:Float, a:Float) = (1, 1, 1, 1)
    var children:[TWGLNode] = []
    
    
    init(name:String, texture:String?, shader:TWGLShaderReference, vertices:[TWGLVertexInfo]?) {
        self.name = name
        self.shader = shader
        if let v = vertices {
            self.vertexCount = v.count
            
            glGenVertexArraysOES(1, &vao)
            glBindVertexArrayOES(vao)


            
            // Generate vertex buffer
            glGenBuffers(1, &vertexBuffer)
            glBindBuffer(GLenum(GL_ARRAY_BUFFER), vertexBuffer)
            glBufferData(GLenum(GL_ARRAY_BUFFER), vertexCount * sizeof(TWGLVertexInfo), v, GLenum(GL_STATIC_DRAW))
            // Enable vertex attributes
            
            let positionSlotFirstComponent: UnsafePointer = UnsafePointer<Int>(bitPattern: 0)
            glEnableVertexAttribArray(0)
            glVertexAttribPointer(0, 3, GLenum(GL_FLOAT), GLboolean(GL_FALSE), GLsizei(sizeof(TWGLVertexInfo)), positionSlotFirstComponent)
            
            let colorSlotFirstComponent: UnsafePointer = UnsafePointer<Int>(bitPattern: sizeof(Float) * 3)
            glEnableVertexAttribArray(TWGLVertexAttrib.Color.rawValue)
            glVertexAttribPointer(TWGLVertexAttrib.Color.rawValue, 4, GLenum( GL_FLOAT), GLboolean(GLboolean(GL_FALSE)), GLsizei(sizeof(TWGLVertexInfo)), colorSlotFirstComponent)
            
            let texCoordSlotFirstComponent: UnsafePointer = UnsafePointer<Int>(bitPattern: sizeof(Float) * 7)
            glEnableVertexAttribArray(GLuint(TWGLVertexAttrib.TexCoord.rawValue))
            glVertexAttribPointer(GLuint(TWGLVertexAttrib.TexCoord.rawValue), 2, GLenum(GL_FLOAT), GLboolean(GL_FALSE), GLsizei(sizeof(TWGLVertexInfo)), texCoordSlotFirstComponent)
            
            let normalSlotFirstComponent: UnsafePointer = UnsafePointer<Int>(bitPattern: sizeof(Float) * 9)
            glEnableVertexAttribArray(GLuint(TWGLVertexAttrib.Normal.rawValue))
            glVertexAttribPointer(GLuint(TWGLVertexAttrib.Normal.rawValue), 3, GLenum(GL_FLOAT), GLboolean(GL_FALSE), GLsizei(sizeof(TWGLVertexInfo)), normalSlotFirstComponent)
            
            glBindVertexArrayOES(0)
            glBindBuffer(GLenum(GL_ARRAY_BUFFER), 0)
            glBindBuffer(GLenum(GL_ELEMENT_ARRAY_BUFFER), 0)
            
            if let t = texture {
                loadTexture(t)
            }
        }
    }
    
    
    func loadTexture(filename:String) {
        let path = NSBundle.mainBundle().pathForResource(filename, ofType: nil)
        
        
        let options = [GLKTextureLoaderOriginBottomLeft : true]
        
        do {
            let info = try GLKTextureLoader.textureWithContentsOfFile(path!, options: options)
            self.texture = info.name
        } catch let error as NSError {
            print("Error loading file: %@", error.localizedDescription)
        }
    }
    
    
    func modelMatrix() -> GLKMatrix4 {
        var modelMatrix = GLKMatrix4Identity
        modelMatrix = GLKMatrix4Translate(modelMatrix, position.x, position.y, position.z)
        modelMatrix = GLKMatrix4Rotate(modelMatrix, rotation.x, 1, 0, 0)
        modelMatrix = GLKMatrix4Rotate(modelMatrix, rotation.y, 0, 1, 0)
        modelMatrix = GLKMatrix4Rotate(modelMatrix, rotation.z, 0, 0, 1)
        modelMatrix = GLKMatrix4Scale(modelMatrix, scale, scale, scale)
        return modelMatrix
    }
    
    var savedModelViewMatrix = GLKMatrix4Identity
    
    func render(parentModelViewMatrix:GLKMatrix4) {
        savedModelViewMatrix = GLKMatrix4Multiply(parentModelViewMatrix, modelMatrix())
        children.forEach { $0.render(savedModelViewMatrix) }
        
        shader!.modelViewMatrix = savedModelViewMatrix
        shader!.texture = texture
        shader!.matColor = GLKVector4Make(matColor.r, matColor.g, matColor.b, matColor.a)
        shader!.prepareToDraw()
        
        glBindVertexArrayOES(vao)
        glDrawArrays(GLenum(GL_TRIANGLES), 0, Int32(vertexCount))
        glBindVertexArrayOES(0)
    }
    
    func update(delta:NSTimeInterval) {
        children.forEach { $0.update(delta) }
    }
    
    
    //    - (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    //    }
    //    - (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event {
    //    }
    //    - (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
    //    }
    
    //    - (CGRect)boundingBoxWithModelViewMatrix:(GLKMatrix4)parentModelViewMatrix {
    //
    //    GLKMatrix4 modelViewMatrix = GLKMatrix4Multiply(parentModelViewMatrix, [self modelMatrix])
    //
    //    GLKVector4 lowerLeft = GLKVector4Make(-self.width/2, -self.height/2, 0, 1)
    //    lowerLeft = GLKMatrix4MultiplyVector4(modelViewMatrix, lowerLeft)
    //    GLKVector4 upperRight = GLKVector4Make(self.width/2, self.height/2, 0, 1)
    //    upperRight = GLKMatrix4MultiplyVector4(modelViewMatrix, upperRight)
    //
    //    CGRect boundingBox = CGRectMake(lowerLeft.x, lowerLeft.y, upperRight.x - lowerLeft.x, upperRight.y - lowerLeft.y)
    //    return boundingBox
    
}